npm error

As we can see in https://github.com/cockpit-project/cockpit-ostree/pull/31#issuecomment-731598411, the npm fails due to the permission of the `.npm` directory.

What about rebuilding the image?

Or, doing that `chown` command during startup of the container?

> What about rebuilding the image?
> Or, doing that `chown` command during startup of the container?

I think neither of those would work.

```
$ docker run -ti --rm -u 123456 usercont/sandcastle bash
$ npm install xyz
npm ERR!   sudo chown -R 123456:0 "/.npm"
$ ls -la /.npm
ls: cannot access '/.npm': No such file or directory
$ echo $HOME
/
```

The `npm` command tries to access `$HOME/.npm`, but given the `$HOME` is `/` it accesses `/.npm`.

After adding `RUN mkdir /.npm && chmod a+rwx /.npm/` into Dockerfile I saw
`npm ERR! Error: EACCES: permission denied, access '/src'`
and after adding `RUN chmod a+rwx /src/` I got
`npm update check failed, try sudo chown -R $USER:$(id -gn $USER) /.config`
so that's probably not the way and I think we need something like [this](https://github.com/packit/packit-service/blob/master/files/setup_env_in_openshift.sh) when we start the container.
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,18 +1,12 @@
 # This is the default sandbox container image to run untrusted commands.
 # It should contain basic utilities, such as git, make, rpmbuild, etc.
-FROM fedora:31
+FROM docker.io/usercont/base
 
-# ANSIBLE_STDOUT_CALLBACK - nicer output from the playbook run
-ENV LANG=en_US.UTF-8 \
-    LC_ALL=C \
+ENV LC_ALL=C \
     PYTHONDONTWRITEBYTECODE=yes \
-    WORKDIR=/src \
-    ANSIBLE_PYTHON_INTERPRETER=/usr/bin/python3 \
-    ANSIBLE_STDOUT_CALLBACK=debug
+    WORKDIR=/src
 
 WORKDIR ${WORKDIR}
-
-RUN dnf install -y ansible
 
 COPY files/install-rpm-packages.yaml /src/files/install-rpm-packages.yaml
 RUN ansible-playbook -vv -c local -t basic-image -i localhost, files/install-rpm-packages.yaml \

--- a/Dockerfile
+++ b/Dockerfile
@@ -4,14 +4,17 @@ FROM docker.io/usercont/base
 
 ENV LC_ALL=C \
     PYTHONDONTWRITEBYTECODE=yes \
-    WORKDIR=/src
+    USER=sandcastle \
+    HOME=/home/sandcastle
 
-WORKDIR ${WORKDIR}
+WORKDIR ${HOME}
+# So the arbitrary user ID can access it.
+RUN chmod g+rwX .
 
-COPY files/install-rpm-packages.yaml /src/files/install-rpm-packages.yaml
-RUN ansible-playbook -vv -c local -t basic-image -i localhost, files/install-rpm-packages.yaml \
+COPY files/install-rpm-packages.yaml ./
+RUN ansible-playbook -vv -c local -t basic-image -i localhost, install-rpm-packages.yaml \
     && dnf clean all
 
-COPY files/container-cmd.sh /src/
+COPY files/container-cmd.sh files/setup_env_in_openshift.sh ./
 # default command is sleep - so users can do .exec(command=[...])
-CMD ["/src/container-cmd.sh"]
+CMD ["./container-cmd.sh"]

--- a/Dockerfile.tests
+++ b/Dockerfile.tests
@@ -2,11 +2,10 @@
 # The expectation is to create sandbox pods from this one.
 FROM docker.io/usercont/sandcastle
 
-COPY tests/requirements.txt /src/tests/requirements.txt
-# Install packages first and reuse the cache as much as possible
-RUN cd /src/ \
-    && ansible-playbook -vv -c local -t with-sandcastle-deps -i localhost, files/install-rpm-packages.yaml
+COPY tests/requirements.txt ./tests/
+RUN ansible-playbook -vv -c local -t with-sandcastle-deps -i localhost, install-rpm-packages.yaml \
+    && dnf clean all
 
-COPY . /src/
+COPY . ./
 
 CMD ["make", "check"]

--- a/files/container-cmd.sh
+++ b/files/container-cmd.sh
@@ -1,4 +1,8 @@
 #!/bin/bash
 
+set -x
+
+source ./setup_env_in_openshift.sh
+
 # 10 minutes is a default timeout for the sandbox pod
 sleep 600

--- a/files/install-rpm-packages.yaml
+++ b/files/install-rpm-packages.yaml
@@ -44,6 +44,7 @@
           - libarchive-devel
           - rpm-devel
           - audit-libs-devel
+          - nss_wrapper
         state: present
       tags:
         - basic-image

--- a/files/install-rpm-packages.yaml
+++ b/files/install-rpm-packages.yaml
@@ -45,7 +45,6 @@
           - rpm-devel
           - audit-libs-devel
         state: present
-        install_weak_deps: no # on F30 python2 pulls in unwanted python-unversioned-command
       tags:
         - basic-image
     - name: Install all RPM packages needed to hack on sandcastle.

--- a/files/setup_env_in_openshift.sh
+++ b/files/setup_env_in_openshift.sh
@@ -1,0 +1,10 @@
+#!/usr/bin/bash
+
+set -x
+
+# Generate passwd file based on current uid, needed for fedpkg
+grep -v ^sandcastle /etc/passwd > "${HOME}/passwd"
+printf "sandcastle:x:$(id -u):0:Sandcastle:${HOME}:/bin/bash\n" >> "${HOME}/passwd"
+export LD_PRELOAD=libnss_wrapper.so
+export NSS_WRAPPER_PASSWD="${HOME}/passwd"
+export NSS_WRAPPER_GROUP=/etc/group

--- a/sandcastle/api.py
+++ b/sandcastle/api.py
@@ -263,12 +263,7 @@ class Sandcastle(object):
             logger.info("loading kubeconfig")
             config.load_kube_config(client_configuration=configuration)
         if not configuration.api_key:
-            raise SandcastleException(
-                f"We loaded the kube config but can't access any cluster. "
-                "We have this configutation:\n"
-                f"Auth: {configuration.auth_settings()}\n"
-                f"Host: {configuration.get_host_settings()}\n"
-            )
+            raise SandcastleException("No api_key, can't access any cluster.\n")
         # example exec.py sets this:
         # https://github.com/kubernetes-client/python/blob/master/examples/exec.py#L61
         configuration.assert_hostname = False

--- a/sandcastle/api.py
+++ b/sandcastle/api.py
@@ -255,14 +255,20 @@ class Sandcastle(object):
         load service account identity, otherwise load kubeconfig
         """
         logger.debug("Initialize kubernetes client")
+        configuration = client.Configuration()
         if "KUBERNETES_SERVICE_HOST" in os.environ:
             logger.info("loading incluster config")
-            config.load_incluster_config()
+            config.load_incluster_config(client_configuration=configuration)
         else:
             logger.info("loading kubeconfig")
-            config.load_kube_config()
-        configuration = client.Configuration()
-        assert configuration.api_key
+            config.load_kube_config(client_configuration=configuration)
+        if not configuration.api_key:
+            raise SandcastleException(
+                f"We loaded the kube config but can't access any cluster. "
+                "We have this configutation:\n"
+                f"Auth: {configuration.auth_settings()}\n"
+                f"Host: {configuration.get_host_settings()}\n"
+            )
         # example exec.py sets this:
         # https://github.com/kubernetes-client/python/blob/master/examples/exec.py#L61
         configuration.assert_hostname = False

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -21,13 +21,15 @@
 # SOFTWARE.
 import time
 from pathlib import Path
+from subprocess import check_output
 from typing import Optional, Any, Dict
 
 import pytest
 import urllib3
-from kubernetes import config, client
+from kubernetes import client
 from kubernetes.client import V1DeleteOptions
 from kubernetes.client.rest import ApiException
+from kubernetes.config import new_client_from_config
 
 from sandcastle.api import Sandcastle
 from sandcastle.utils import run_command, get_timestamp_now, clean_string
@@ -35,7 +37,8 @@ from sandcastle.utils import run_command, get_timestamp_now, clean_string
 NON_EX_IMAGE = "non-ex-image"
 PROJECT_NAME = "cyborg"
 SANDBOX_IMAGE = "docker.io/usercont/sandcastle"
-TEST_IMAGE_NAME = "docker.io/usercont/sandcastle-tests"
+TEST_IMAGE_BASENAME = "sandcastle-tests"
+TEST_IMAGE_NAME = f"docker.io/usercont/{TEST_IMAGE_BASENAME}"
 NAMESPACE = "myproject"
 SANDCASTLE_MOUNTPOINT = "/sandcastle"
 
@@ -66,6 +69,11 @@ def build_now():
     )
 
 
+def enable_user_access_namespace(user: str, namespace: str):
+    c = ["oc", "adm", "-n", namespace, "policy", "add-role-to-user", "edit", user]
+    run_command(c)
+
+
 # TODO: refactor into a class
 def run_test_within_pod(
     test_path: str, with_pv_at: Optional[str] = None, new_namespace: bool = False
@@ -77,21 +85,34 @@ def run_test_within_pod(
     :param with_pv_at: path to PV within the pod
     :param new_namespace: create new namespace and pass it via env var
     """
-    config.load_kube_config()
-    configuration = client.Configuration()
-    assert configuration.api_key
-    api = client.CoreV1Api(client.ApiClient(configuration))
+    # this will connect to the cluster you have active right now - see `oc status`
+    current_context = check_output(["oc", "config", "current-context"]).strip().decode()
+    api_client = new_client_from_config(context=current_context)
+    api = client.CoreV1Api(api_client)
+
+    # you need this for minishift; or do `eval $(minishift docker-env) && make build`
+    # # FIXME: get this from env or cli (`minishift openshift registry`)
+    # also it doesn't work for me right now, unable to reach minishift's docker registry
+    # registry = "172.30.1.1:5000"
+    # openshift_image_name = f"{registry}/myproject/{TEST_IMAGE_BASENAME}"
+    # # {'authorization': 'Bearer pRW5rGmqgBREnCeVweLcHbEhXvluvG1cImWfIrxWJ2A'}
+    # api_token = list(api_client.configuration.api_key.values())[0].split(" ")[1]
+    # check_call(["docker", "login", "-u", "developer", "-p", api_token, registry])
+    # check_call(["docker", "tag", TEST_IMAGE_NAME, openshift_image_name])
+    # check_call(["docker", "push", openshift_image_name])
 
     pod_name = f"test-orchestrator-{get_timestamp_now()}"
 
     cont_cmd = [
         "bash",
         "-c",
-        "ls -lha " f"&& pytest-3 -vv -l -p no:cacheprovider {test_path}",
+        "~/setup_env_in_openshift.sh "
+        "&& ls -lha && pwd && id "
+        f"&& pytest-3 -vv -l -p no:cacheprovider {test_path}",
     ]
 
     container: Dict[str, Any] = {
-        "image": TEST_IMAGE_NAME,
+        "image": TEST_IMAGE_NAME,  # make sure this is correct
         "name": pod_name,
         "tty": True,  # corols
         "command": cont_cmd,
@@ -99,22 +120,14 @@ def run_test_within_pod(
         "env": [],
     }
 
+    user = f"system:serviceaccount:{NAMESPACE}:default"
+    enable_user_access_namespace(user, NAMESPACE)
     test_namespace = None
     if new_namespace:
         test_namespace = f"sandcastle-tests-{get_timestamp_now()}"
         c = ["oc", "new-project", test_namespace]
         run_command(c)
-        c = [
-            "oc",
-            "adm",
-            "-n",
-            test_namespace,
-            "policy",
-            "add-role-to-user",
-            "edit",
-            f"system:serviceaccount:{NAMESPACE}:default",
-        ]
-        run_command(c)
+        enable_user_access_namespace(user, test_namespace)
         container["env"] += [
             {"name": "SANDCASTLE_TESTS_NAMESPACE", "value": test_namespace}
         ]

